### PR TITLE
Adding off property to slider.prototype

### PR DESCRIPTION
**Pull request by arjo1203**
Just like the 'on' property attaches an eventHandlers to a slider. The 'off' property gives the option to remove eventHandlers from slider.

### CODE REVIEW

**Comment by seiyria** [comment]
Seems fine. I think your indentation is a bit whack, though. It'd also be good to have some mirrored tests [here](https://github.com/seiyria/bootstrap-slider/blob/master/test/specs/EventsSpec.js) to account for your changes.

**Comment by arjo1203** [comment]
Thanks for the feedback.

**Comment by seiyria** [comment]
You'll also want to make sure your code passes the build, too.

**Comment by rovolution** [comment]
I would like to see automated tests + documentation for the new public method

**Comment by arjo1203** [comment]
I am new to testing, could you give me some pointers or resources. I was looking at Jasmine but I am confused how to test my 'off'  method.

File: js/bootstrap-slider.js [hunk 1]
```diff
@@ -893,6 +902,15 @@
 				}
 				this.eventToCallbackMap[evt].push(callback);
 			},
+            _unbindNonQueryEventHandler: function(evt, callback) {
+                if(this.eventToCallbackMap[evt]!==undefined) {
+                    for (var i = 0; i < this.eventToCallbackMap[evt].length(); i++) {
+                        if (this.eventToCallbackMap[evt][i] == callback) {
+                            this.eventToCallbackMap[evt].splice(i, 1);
```
Comment (rovolution) [inline js/bootstrap-slider.js]: i would add a `break;` statement after this line, which will prevent the code from looping through all of the event to callback mappings;

File: js/bootstrap-slider.js [hunk 2]
```diff
@@ -893,6 +902,15 @@
 				}
 				this.eventToCallbackMap[evt].push(callback);
 			},
+            _unbindNonQueryEventHandler: function(evt, callback) {
+                if(this.eventToCallbackMap[evt]!==undefined) {
+                    for (var i = 0; i < this.eventToCallbackMap[evt].length(); i++) {
+                        if (this.eventToCallbackMap[evt][i] == callback) {
```
Comment (rovolution) [inline js/bootstrap-slider.js]: can you use === instead of ==?

File: js/bootstrap-slider.js [hunk 3]
```diff
@@ -893,6 +902,15 @@
 				}
 				this.eventToCallbackMap[evt].push(callback);
 			},
+            _unbindNonQueryEventHandler: function(evt, callback) {
+                if(this.eventToCallbackMap[evt]!==undefined) {
```
Comment (rovolution) [inline js/bootstrap-slider.js]: fix the spacing on this

```js
this.eventToCallbackMap[evt] !== undefined
```

**Comment by rovolution** [comment]
What I would do is write a test where you do the following:

1) Create a slider and bind to an event
2) Inside the event handler, have it set a variable (probably called `eventHandlerTriggered`) to `true`. Have the variable initially be set to `false`
3) Use JavaScript to simulate a `mousedown` event on the slider
4) Check to see if the value `eventHandlerTriggered` is still `false`. if it is, then your test is successful (the event listener was removed). If not, your test failed and you need to double check your implementation

check out the files in our [/test/specs](https://github.com/seiyria/bootstrap-slider/tree/master/test/specs) directory for examples. Create a separate spec file for this

**Comment by arjo1203** [comment]
Thank you for the feedback. I'll work on it and get back to you.

**Comment by rovolution** [comment]
cool thanks!!!

**Comment by arjo1203** [comment]
Where is your 'testSlider1' reference? Do I need to make another slider to test with?

**Comment by rovolution** [comment]
@arjo1203 This is how the test suite works:

1) Compile the /tpl/SpecRunner.tpl into an HTML file
2) load that file into a headless browser (PhantomJS)
3) Run the test suite in the headless browser with Jasmine
4) Output the test results in the console
5) delete the temporary SpecRunner.html file

So if you look in the /tpl/SpecRunner.tpl file, you will see the references to the testSlider element. Does that make more sense?

**Comment by arjo1203** [comment]
I think I was able to figure it. Still not completely sure but I was able to have my test suite pass and 
the console output was:

151 specs in 0.702s.
>> 0 failures

Running "template:generate-index-page" (template) task
File `index.html` created.

Done, without errors.

**Comment by rovolution** [comment]
pretty solid test! good job!

1) Rename the file from offSpec.js -> OffMethodSpec.js
2) Add some documentation for the method to the README.md file under the `Functions` section

File: test/specs/offSpec.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    };
+
+
+    beforeEach(function() {
+        eventHandlerTriggered = false;
+        mouse = document.createEvent('MouseEvents');
+    });
+
+
+    it("'slideStart' event is triggered properly and can be binded to", function() {
```
Comment (rovolution) [inline test/specs/offSpec.js]: change this to `it("should properly unbind an event listener")`

**Comment by arjo1203** [comment]
Is there anything I need to do??

**Comment by seiyria** [comment]
It looks fine to me after your latest changes!

File: js/bootstrap-slider.js [hunk 5]
```diff
@@ -888,11 +897,21 @@
...
 			},
 			_bindNonQueryEventHandler: function(evt, callback) {
-				if(this.eventToCallbackMap[evt]===undefined) {
+				if(this.eventToCallbackMap[evt] === undefined) {
 					this.eventToCallbackMap[evt] = [];
 				}
 				this.eventToCallbackMap[evt].push(callback);
 			},
+            _unbindNonQueryEventHandler: function(evt, callback) {
+                if(this.eventToCallbackMap[evt] !== undefined) {
```
Comment (rovolution) [inline js/bootstrap-slider.js]: at the top of this function, I would put `this.eventToCallbackMap[evt]` into a variable and then reference that throughout the function. 

```js
var boundCallbacks = this.eventToCallbackMap[evt];
```

**Comment by rovolution** [comment]
@arjo1203 left some minor feedback, but after that it looks good. great work! tag me once you make the update and ill merge/bump the version/publish to NPM

**Comment by arjo1203** [comment]
I believe our work is done!

**Comment by rovolution** [comment]
done! thanks for your contribution!

**Comment by arjo1203** [comment]
Just wanted to thank you guys for the help and if I could ever help out with anything I am just a message away. Also I am not sure why the indentation is so off for the extra method I add slider.prototype, on my local file the indentation is prefect.

**Pull request merged**